### PR TITLE
Move copy context to policy

### PR DESCRIPTION
**Pull request by qinsoon**
This PR is a step towards https://github.com/mmtk/mmtk-core/issues/258 and https://github.com/mmtk/mmtk-core/issues/110. It mainly removes the plan-specific copy context, and replaces with a configurable GC worker copy context that uses policy-specific copy context (copy allocators). The `GCWorkerCopyContext` works similarly to the configurable `Mutator`:
* `GCWorkerCopyContext` ≈ `Mutator`
* `CopySemantics` ≈ `AllocationSemantics`
* `CopyConfig` ≈ `MutatorConfig`

This PR
* renames `CopyContext` (plan-specific) to `PolicyCopyContext`, and moves it to `policy/copy_context`.
  * only copy policies provide implementation for it, namely `CopySpaceCopyContext` and `ImmixCopyContext`.
* adds a `GCWorkerCopyContext` which is a combination of `PolicyCopyContext`, and will be used as the thread local data structure for a GC worker.
  * `CopySemantics` is introduced.
  * A plan needs to provide a `CopyConfig` that maps `CopySemantics` to copy allocators, and map copy allocators to spaces.
  * For copy operation, a proper `CopySemantics` need to be provided.
  * `GCWorkerLocalPtr` and a few type parameters about the `CopyContext` is removed, as now GC worker local and the copy context has the fixed type of `GCWorkerCopyContext`.
* removes the plan specific copy context (such as `SSCopyContext`, `GenCopyCopyContext`, etc)

Related binding PRs:
* https://github.com/mmtk/mmtk-openjdk/pull/128
* https://github.com/mmtk/mmtk-jikesrvm/pull/93
* https://github.com/mmtk/mmtk-v8/pull/47

### CODE REVIEW

File: src/policy/copyspace.rs [hunk 1]
```diff
@@ -215,6 +217,7 @@ impl<VM: VMBinding> CopySpace<VM> {
             new_object
         } else {
             trace!("... no it isn't. Copying");
+            let copy_context = unsafe { worker.local::<CopySpaceCopyContext<VM>>() };
             let new_object =
                 object_forwarding::forward_object::<VM, _>(object, semantics, copy_context);
```
Comment (wks) [inline src/policy/copyspace.rs:222]: `copy_context` contains information about where the object is copied **to**.  For example, if this CopySpace is used as the nursery of the GenImmix plan, then the `copy_context` here should be `ImmixCopyContext`, not `CopySpaceCopyContext`.

Comment (qinsoon) [reply src/policy/copyspace.rs:222]: Yeah. Exactly. I am working on this. This PR is not yet ready for review.

**Comment by qinsoon** [comment]
openjdk-binding-test
OPENJDK_BINDING_REF=update-pr-499

**Comment by qinsoon** [comment]
jikesrvm-binding-test
JIKESRVM_BINDING_REF=update-pr-499

**Comment by qinsoon** [comment]
v8-binding-test
V8_BINDING_REF=update-pr-499

File: src/scheduler/worker.rs [hunk 2]
```diff
@@ -126,15 +85,14 @@ impl<VM: VMBinding> GCWorker<VM> {
...
-    pub unsafe fn local<W: 'static + GCWorkerLocal>(&mut self) -> &mut W {
-        self.local.as_type::<W>()
+    // TODO: We should be able to remove this method. We can create the copy context without a proper tls.
+    // In init(), we set tls for the worker and for the copy context.
+    pub fn set_local(&mut self, copy: GCWorkerCopyContext<VM>) {
+        self.copy = copy;
     }
 
-    pub fn set_local(&mut self, local: GCWorkerLocalPtr) {
-        self.local = local;
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply src/scheduler/worker.rs:137]: I can do this change in a separate PR. If you think this should be done in this PR, I can do it as well. Please let me know.

File: src/plan/global.rs [hunk 3]
```diff
@@ -203,11 +123,12 @@ pub trait Plan: 'static + Sync + Downcast {
...
-        &self,
-        tls: VMWorkerThread,
-        mmtk: &'static MMTK<Self::VM>,
-    ) -> GCWorkerLocalPtr;
+
+    /// Create thread local GC worker. For copying plan, they will have to override this method,
+    /// and use the correct copy context as GC workers.
+    fn create_worker_local(&'static self, _tls: VMWorkerThread) -> GCWorkerCopyContext<Self::VM> {
+        GCWorkerCopyContext::<Self::VM>::new_non_copy()
+    }
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply src/plan/global.rs:131]: Sounds good. I will make the change.

[the remark this answers is not in the document]
Comment (qinsoon) [reply src/plan/global.rs]: In the end, I made `create_worker()` as a static function. I cannot put it in the trait for a compile error. In L129, `self` is used as `&dyn Plan`, but `self` has an unknown size (as it is in the trait's default implementation). Previously, `create_worker_local()` was implemented by each plan, and `self` has a know size, so it was fine.

File: src/scheduler/worker.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wks) [reply src/scheduler/worker.rs:137]: It's not that bad. We can postpone that to another PR.

**Comment by qinsoon** [comment]
@wks Do you have any other comments? Can you review the binding PRs as well (those can be found in the PR description)?

**Review by wks — APPROVED** [review]
LGTM.

**Pull request merged**